frontend/app/LoginRefreshComponent.tsx

```diff
@@ -0,0 +1,110 @@
+import React, {useState, useEffect, useRef} from "react";
+import {JwtDataShape} from "./DecodedProfile";
+import {refreshLogin} from "./OAuth2Helper";
+
+interface LoginComponentProps {
+    refreshToken?: string;
+    checkInterval?:number;
+    loginData: JwtDataShape;
+    onLoginRefreshed?: ()=>void;
+    onLoginCantRefresh?: (reason:string)=>void;
+    onLoginExpired: ()=>void;
+    onLoggedOut?: ()=>void;
+    overrideRefreshLogin?: (tokenUri:string)=>Promise<void>;    //only used for testing
+    tokenUri: string;
+}
+
+const LoginRefreshComponent:React.FC<LoginComponentProps> = (props) => {
+    const [refreshFailed, setRefreshFailed] = useState<boolean>(false);
+
+    let loginDataRef = useRef(props.loginData);
+    const tokenUriRef = useRef(props.tokenUri);
+    const overrideRefreshLoginRef = useRef(props.overrideRefreshLogin);
+
+    useEffect(()=>{
+        const intervalTimerId = window.setInterval(checkExpiryHandler, props.checkInterval ?? 60000);
+
+        return (()=>{
+            console.log("Removing checkExpiryHandler")
+            window.clearInterval(intervalTimerId);
+        })
+    }, []);
+
+    useEffect(()=>{
+        console.log("refreshFailed was toggled to ", refreshFailed);
+        if(refreshFailed) {
+            console.log("Setting countdown handler");
+            const intervalTimerId = window.setInterval(updateCountdownHandler, 1000);
+            return (()=>{
+                console.log("Cleared countdown handler");
+                window.clearInterval(intervalTimerId);
+            })
+        }
+    }, [refreshFailed]);
+
+    useEffect(()=>{
+        loginDataRef.current = props.loginData;
+    }, [props.loginData]);
+
+    /**
+     * Called periodically every second once a refresh has failed to alert the user how long they have left
+     */
+    const updateCountdownHandler = () => {
+        const nowTime = new Date().getTime() / 1000; //assume time is in seconds
+        const expiry = loginDataRef.current.exp;
+        const timeToGo = expiry - nowTime;
+
+        if(timeToGo>1) {
```

Comment:
could be simplified to `if(timeToGo<=1 && props.onLoginExpired) props.onLoginExpired();`

Reply:
Done.